### PR TITLE
Adds test for Ueberauth.Strategy.Slack.OAuth#get_token!

### PR DESCRIPTION
**Pull request by alvincrespo**
I'm currently looking to use this library for a side project. In doing so, I'm writing up tests to better understand how it is developed. This is my first attempt, and it has to do with creating a test against the `Ueberauth.Strategy.Slack.OAuth#get_token!` function.

## Changes

- Creates a config based attribute in the Ueberauth.Strategy.Slack.OAuth module
  - This attribute will be used to determine which client to use in each environment (dev, test, prod, etc...)
- Defines `config :ueberauth, :oauth2_client, Ueberauth.Client.OAuth2` as the config for test
- Defines `config :ueberauth, :oauth2_client, OAuth2.Client` for all other environments
- Adds `config :ueberauth, Ueberauth.Strategy.Slack.OAuth,` for test environment
- Fixes some warnings from testing (e.g. Dict.get has been deprecated)
- Adds `{:poison, "~> 2.0", only: :test}` for testing
- Creates two tests for checking against invalid/valid codes to the Slack API and their expected return values

### References

http://blog.plataformatec.com.br/2015/10/mocks-and-explicit-contracts/

### CODE REVIEW

File: .tool-versions [hunk 1]
```diff
@@ -0,0 +1 @@
+elixir 1.6.1
```
Comment (alvincrespo) [inline .tool-versions]: I'm using asdf to manage my elixir version, I added this by mistake - but not sure if we want to specify an explicit version of elixir to use?

Comment (doomspork) [reply .tool-versions]: Let's not add these files, they tend to cause unnecessary churn 👌

File: lib/ueberauth/client/oauth2.ex [hunk 2]
```diff
@@ -0,0 +1,26 @@
+defmodule Ueberauth.Client.OAuth2 do
+  def get_token!(_, params \\ [], _ \\ [], _ \\ []) do
+    [code: code] = params
+
+    case code  do
+      "abc123" -> %{
```
Comment (doomspork) [inline lib/ueberauth/client/oauth2.ex]: These are the actual two codes returned by Slack?

File: lib/ueberauth/client/oauth2.ex [hunk 3]
```diff
@@ -0,0 +1,26 @@
+defmodule Ueberauth.Client.OAuth2 do
+  def get_token!(_, params \\ [], _ \\ [], _ \\ []) do
+    [code: code] = params
+
+    case code  do
```
Comment (doomspork) [inline lib/ueberauth/client/oauth2.ex]: Extra whitespace here.

File: lib/ueberauth/client/oauth2.ex [hunk 4]
```diff
@@ -0,0 +1,26 @@
+defmodule Ueberauth.Client.OAuth2 do
+  def get_token!(_, params \\ [], _ \\ [], _ \\ []) do
+    [code: code] = params
```
Comment (doomspork) [inline lib/ueberauth/client/oauth2.ex]: You could probably move this to the function head couldn't you?

File: lib/ueberauth/strategy/slack/oauth.ex [hunk 5]
```diff
@@ -34,11 +37,14 @@ defmodule Ueberauth.Strategy.Slack.OAuth do
   end
 
   def get_token!(params \\ [], options \\ %{}) do
-    headers        = Dict.get(options, :headers, [])
-    options        = Dict.get(options, :options, [])
-    client_options = Dict.get(options, :client_options, [])
-    client = OAuth2.Client.get_token!(client(client_options), params, headers, options)
-    client.token
+    %{headers: headers, options: %{ options: opts } } = options
```
Comment (doomspork) [inline lib/ueberauth/strategy/slack/oauth.ex]: Can you please remove extra whitespace around the `{`, we want to try to be consistent 👍

File: lib/ueberauth/strategy/slack/oauth.ex [hunk 6]
```diff
@@ -34,11 +37,14 @@ defmodule Ueberauth.Strategy.Slack.OAuth do
...
+    [client_options: client_options] = opts
+
+    created_client = client_options |> client()
```
Comment (doomspork) [inline lib/ueberauth/strategy/slack/oauth.ex]: On this like and the next please don't use pipes if it's only a single function call.

**Review by doomspork — CHANGES_REQUESTED** [review]
Thank you for this work @alvincrespo, I've got some questions for you.

File: lib/ueberauth/client/oauth2.ex (hunk 2 above)
Comment (alvincrespo) [reply lib/ueberauth/client/oauth2.ex]: Nope. These are just mock codes that are sent to Slack for them to authenticate.

**Comment by alvincrespo** [comment]
@doomspork Awesome! Thanks for the feedback, just updated.

**Comment by doomspork** [comment]
@alvincrespo can you rebase?  I'll merge 😀

**Comment by alvincrespo** [comment]
@doomspork Thanks so much! Rebased.

**Review by yordis — CHANGES_REQUESTED** [review]
Why do you need a `dev` configuration on this project?

Elixir config only works on the higher order project and this project is meant to be used in other projects so I don't see why we need this `dev` config.

**Review by yordis — CHANGES_REQUESTED** [review]
btw, could add some jump lines at the end of the file of `config.exs` and `dev.exs`?

**Comment by alvincrespo** [comment]
@yordis Sorry, it's been a while since I worked on this - but I think I needed the config/dev.exs file in order to get the test to pick up which configuration to use. Which sounds weird, because I think that would be in config/test.exs right?

Anyways, I can remove it and test it out later and have a better reason if you prefer?

**Comment by yordis** [comment]
@alvincrespo yeah but you don't even need `test.exs` because you are only using this package in `test` mode when you are developing it so you can just put all the config in `config.exs` only.

If tomorrow for some reason we invent some environment like `regression.exs` then we would need to import per environment based but it is not the case.

So, just use `config.exs`

**Comment by alvincrespo** [comment]
@yordis yeah, that makes sense. I might have been floundering about something a while ago - so I removed those and added a new line at the end of config/config.exs (I think that's what you meant by jump line?)

**Comment by yordis** [comment]
@alvincrespo it is fixed now. 

The newline character, sorry.

File: lib/ueberauth/strategy/slack/oauth.ex [hunk 7]
```diff
@@ -34,13 +37,14 @@ defmodule Ueberauth.Strategy.Slack.OAuth do
...
-    headers        = Dict.get(options, :headers, [])
-    options        = Dict.get(options, :options, [])
-    client_options = Dict.get(options, :client_options, [])
+    %{headers: headers, options: %{options: opts}} = options
+    [client_options: client_options] = opts
+
+    created_client = client(client_options)
 
-    client = OAuth2.Client.get_token!(client(client_options), params, headers, options)
+    %{token: token} = @oauth2_client.get_token!(created_client, params, headers, opts)
```
Comment (yordis) [inline lib/ueberauth/strategy/slack/oauth.ex]: You could write it this way and reads better.

```elixir
created_client
|> @oauth2_client.get_token!(params, headers, opts)
|> Map.get(:token)
```

File: lib/ueberauth/strategy/slack/oauth.ex [hunk 8]
```diff
@@ -74,4 +78,4 @@ defmodule Ueberauth.Strategy.Slack.OAuth do
 
     endpoint
   end
-end
+end
```
Comment (yordis) [inline lib/ueberauth/strategy/slack/oauth.ex]: Add a newline here.

File: lib/ueberauth/client/oauth2.ex [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+          access_token: "asdf-1234-klsc",
+          expires_at: nil,
+          other_params: %{"ok" => true},
+          refresh_token: nil,
+          token_type: "Bearer"
+        }
+      }
+    end
+  end
+end
```
Comment (yordis) [inline lib/ueberauth/client/oauth2.ex]: Add newline here.

**Review by yordis — CHANGES_REQUESTED** [review]

File: lib/ueberauth/strategy/slack/oauth.ex [hunk 10]
```diff
@@ -59,7 +63,7 @@ defmodule Ueberauth.Strategy.Slack.OAuth do
   defp endpoint("/" <> _path = endpoint, client), do: client.site <> endpoint
   defp endpoint(endpoint, _client), do: endpoint
 
-  defp to_url(client, endpoint, params \\ nil) do
+  defp to_url(client, endpoint, params) do
```
Comment (yordis) [inline lib/ueberauth/strategy/slack/oauth.ex:66]: Why did you remove the default value here?

Comment (alvincrespo) [reply lib/ueberauth/strategy/slack/oauth.ex:66]: Because the following warning:

```
warning: default arguments in to_url/3 are never used
  lib/ueberauth/strategy/slack/oauth.ex:66
```

Which hints to:

```
      if params do
        endpoint <> "?" <> URI.encode_query(params)
      else
        endpoint
      end
```

Basically, we're already checking params - so no need for a default.

File: lib/ueberauth/client/oauth2.ex (hunk 9 above)
Comment (alvincrespo) [reply lib/ueberauth/client/oauth2.ex]: Added. Not sure why GH isn't collapsing the comment.

File: config/config.exs [hunk 11]
```diff
@@ -26,5 +28,5 @@ use Mix.Config
 # by uncommenting the line below and defining dev.exs, test.exs and such.
 # Configuration from the imported file will override the ones defined
 # here (which is why it is important to import them last).
-#
-#     import_config "#{Mix.env}.exs"
+
+import_config "#{Mix.env}.exs"
```
Comment (yordis) [inline config/config.exs]: You removed the environment files so reverse this back

**Review by yordis — CHANGES_REQUESTED** [review]

**Comment by alvincrespo** [comment]
@yordis I commented out the environment config import and now tests fail, which is why I uncommented and created the other environment config files initially:

```
➜  ueberauth_slack git:(ac-add-test-for-get-token) ✗ mix test


  1) test get_token! - an invalid code returns a nil access token (UeberauthSlackTest)
     test/ueberauth_slack_test.exs:9
     ** (FunctionClauseError) no function clause matching in Keyword.merge/2

     The following arguments were given to Keyword.merge/2:

         # 1
         [strategy: Ueberauth.Strategy.Slack.OAuth, site: "https://slack.com/api", authorize_url: "https://slack.com/oauth/authorize", token_url: "https://slack.com/api/oauth.access"]

         # 2
         nil

     code: %{access_token: token} = OAuth.get_token!(params, options)
     stacktrace:
       (elixir) lib/keyword.ex:673: Keyword.merge/2
       (ueberauth_slack) lib/ueberauth/strategy/slack/oauth.ex:19: Ueberauth.Strategy.Slack.OAuth.client/1
       (ueberauth_slack) lib/ueberauth/strategy/slack/oauth.ex:43: Ueberauth.Strategy.Slack.OAuth.get_token!/2
       test/ueberauth_slack_test.exs:19: (test)



  2) test get_token! - a valid code returns a non-nil access token (UeberauthSlackTest)
     test/ueberauth_slack_test.exs:24
     ** (FunctionClauseError) no function clause matching in Keyword.merge/2

     The following arguments were given to Keyword.merge/2:

         # 1
         [strategy: Ueberauth.Strategy.Slack.OAuth, site: "https://slack.com/api", authorize_url: "https://slack.com/oauth/authorize", token_url: "https://slack.com/api/oauth.access"]

         # 2
         nil

     code: %{access_token: token} = OAuth.get_token!(params, options)
     stacktrace:
       (elixir) lib/keyword.ex:673: Keyword.merge/2
       (ueberauth_slack) lib/ueberauth/strategy/slack/oauth.ex:19: Ueberauth.Strategy.Slack.OAuth.client/1
       (ueberauth_slack) lib/ueberauth/strategy/slack/oauth.ex:43: Ueberauth.Strategy.Slack.OAuth.get_token!/2
       test/ueberauth_slack_test.exs:34: (test)



Finished in 0.06 seconds
2 tests, 2 failures

Randomized with seed 310497
```

If we were to have this automatically tested in a CI, we'd get the ☝️ errors.

**Comment by yordis** [comment]
@alvincrespo because you needed the information that were on `text.exs` on `config.exs` but you don't need separate files.

**Comment by doomspork** [comment]
@alvincrespo I'd like to get this merged in, could you rebase it or would you like me to?

**Comment by doomspork** [comment]
It's been more than 2 years. I'm closing this.
